File: 2020-04/4-09/3无重复字符的最长子串/3.py
#  -*- coding: utf-8 -*-

#  Editor      : Pycharm
#  File        : 3.py
#  Author      : ZhenyuMa
#  Created     : 2020/4/9 9:07 下午
#  Description : 滑动窗口的题


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:

        res = 0
        record = {}
        l = 0
        r = -1

        while l < len(s):
            if r + 1 < len(s) and s[r + 1] not in record:
                r += 1
                record[s[r]] = r
            else:
                record.pop(s[l])  # 不满足条件会一直删的，没问题
                l += 1
            res = max(res, r - l + 1)

        return res

File: 2020-05/5-02/3无重复字符的最长子串/3.cpp

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.size() == 0)
            return 0;

        unordered_map<char, int> record;
        int l = 0, r = 0;  // 左闭右开
        int res = 0;

        while(r < s.size()){
            char c = s[r++];  // 在里面加了，所以 <size()是没问题的，最后会让r==s.size()
            record[c]++;

            while(record[c] > 1){
                record[s[l++]]--;
            }

            res = max(res, r - l);
        }
        return res;
    }
};